src/collective/communicator_factory.cc

```diff
@@ -0,0 +1,61 @@
+/*!
+ * Copyright 2022 XGBoost contributors
+ */
+#include "communicator_factory.h"
+
+#include "rabit_communicator.h"
+
+#if defined(XGBOOST_USE_FEDERATED)
+#include "../../plugin/federated/federated_communicator.h"
+#endif
+
+namespace xgboost {
+namespace collective {
+
+#ifndef XGBOOST_USE_CUDA
+thread_local std::unique_ptr<CommunicatorFactory> CommunicatorFactory::instance_{};
+
+CommunicatorFactory::CommunicatorFactory(CommunicatorType type, Communicator* communicator)
+    : type_{type}, communicator_{communicator} {}
+
+void CommunicatorFactory::Init(int argc, char* argv[]) {
+  if (instance_) {
+    LOG(FATAL) << "Communicator factory can only be initialized once.";
+  }
+
+  auto type = GetTypeFromEnv();
+  auto const arg = GetTypeFromArgs(argc, argv);
+  if (arg != CommunicatorType::kUnknown) {
+    type = arg;
+  }
+  switch (type) {
+    case CommunicatorType::kRabit: {
+      RabitCommunicatorFactory factory{argc, argv};
+      auto* comm = factory.Create();
+      instance_.reset(new CommunicatorFactory(type, comm));
```

Comment:
So `CommunicatorFactory` is a factory of factories?

Reply:
Got rid of the factories.